src/map.rs

```diff
@@ -12,50 +12,64 @@ use crate::iter_helpers::{deserialize_kv, deserialize_v};
 #[cfg(feature = "iterator")]
 use crate::keys::Prefixer;
 use crate::keys::{Key, PrimaryKey};
+use crate::namespace::Ns;
 use crate::path::Path;
 #[cfg(feature = "iterator")]
 use crate::prefix::{namespaced_prefix_range, Prefix};
+use crate::Namespace;
 #[cfg(feature = "iterator")]
 use cosmwasm_std::Order;
 use cosmwasm_std::{from_json, Addr, CustomQuery, QuerierWrapper, StdError, StdResult, Storage};
 
 #[derive(Debug, Clone)]
-pub struct Map<'a, K, T> {
-    namespace: &'a [u8],
+pub struct Map<K, T> {
+    namespace: Ns,
     // see https://doc.rust-lang.org/std/marker/struct.PhantomData.html#unused-type-parameters for why this is needed
     key_type: PhantomData<K>,
     data_type: PhantomData<T>,
 }
 
-impl<'a, K, T> Map<'a, K, T> {
-    pub const fn new(namespace: &'a str) -> Self {
+impl<K, T> Map<K, T> {
+    /// Creates a new [`Map`] with the given storage key. This is a const fn only suitable
+    /// when you have the storage key in the form of a static string slice.
+    pub const fn new(namespace: &'static str) -> Self {
         Map {
-            namespace: namespace.as_bytes(),
+            namespace: Ns::from_static_str(namespace),
             data_type: PhantomData,
             key_type: PhantomData,
         }
     }
 
-    pub fn namespace(&self) -> &'a [u8] {
-        self.namespace
+    /// Creates a new [`Map`] with the given storage key. Use this if you might need to handle
+    /// a dynamic string. Otherwise, you might prefer [`Map::new`].
+    pub fn new_dyn(namespace: impl Namespace) -> Self {
+        Map {
+            namespace: namespace.namespace(),
+            data_type: PhantomData,
+            key_type: PhantomData,
+        }
+    }
+
+    pub fn namespace(&self) -> &[u8] {
+        self.namespace.as_slice()
     }
```

Comment:
Here we have a bit of naming trouble. "namespace" is now
- a trait
- a ref-or-owned structure
- a byte slice

Reply:
I could rename this method to namespace_bytes since we're already breaking API?